File: CS/DiagramBindToList/Controllers/HomeController.cs
using DevExpress.Web.Mvc;
using DiagramBindToList.Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;

namespace DiagramBindToList.Controllers
{
    public class HomeController : Controller
    {
        public ActionResult Index()
        {
            return View(WorkflowDataProvider.LoadData());
        }

        public ActionResult DiagramUpdate(MVCxDiagramNodeUpdateValues<DiagramBindToList.Models.Node, int> nodeUpdateValues, MVCxDiagramEdgeUpdateValues<DiagramBindToList.Models.Edge, int> edgeUpdateValues)
        {
            foreach (var item in nodeUpdateValues.Update)
            {
                WorkflowDataProvider.UpdateObject(item);
            }
            foreach (var itemKey in nodeUpdateValues.DeleteKeys)
            {
                WorkflowDataProvider.DeleteObject(itemKey);
            }
            foreach (var item in nodeUpdateValues.Insert)
            {
                var key = WorkflowDataProvider.InsertObject(item);
                nodeUpdateValues.MapInsertedItemKey(item, key);
            }
            foreach (var item in edgeUpdateValues.Update)
            {
                WorkflowDataProvider.UpdateConnection(item);
            }
            foreach (var itemKey in edgeUpdateValues.DeleteKeys)
            {
                WorkflowDataProvider.DeleteConnection(itemKey);
            }
            foreach (var item in edgeUpdateValues.Insert)
            {
                var key = WorkflowDataProvider.InsertConnection(item);
                edgeUpdateValues.MapInsertedItemKey(item, key);
            }
            return DiagramExtension.GetBatchUpdateResult(nodeUpdateValues, edgeUpdateValues);
        }
    }
}

File: VB/DiagramBindToListVB/Controllers/HomeController.vb
Imports DevExpress.Web.Mvc

Public Class HomeController
    Inherits System.Web.Mvc.Controller

    Function Index() As ActionResult
        Return View(WorkflowDataProvider.LoadData())
    End Function

    Public Function DiagramUpdate(ByVal nodeUpdateValues As MVCxDiagramNodeUpdateValues(Of Node, Integer), ByVal edgeUpdateValues As MVCxDiagramEdgeUpdateValues(Of Edge, Integer)) As ActionResult
        For Each item In nodeUpdateValues.Update
            WorkflowDataProvider.UpdateObject(item)
        Next

        For Each itemKey In nodeUpdateValues.DeleteKeys
            WorkflowDataProvider.DeleteObject(itemKey)
        Next

        For Each item In nodeUpdateValues.Insert
            Dim key = WorkflowDataProvider.InsertObject(item)
            nodeUpdateValues.MapInsertedItemKey(item, key)
        Next

        For Each item In edgeUpdateValues.Update
            WorkflowDataProvider.UpdateConnection(item)
        Next

        For Each itemKey In edgeUpdateValues.DeleteKeys
            WorkflowDataProvider.DeleteConnection(itemKey)
        Next

        For Each item In edgeUpdateValues.Insert
            Dim key = WorkflowDataProvider.InsertConnection(item)
            edgeUpdateValues.MapInsertedItemKey(item, key)
        Next

        Return DiagramExtension.GetBatchUpdateResult(nodeUpdateValues, edgeUpdateValues)
    End Function
End Class